internal/core/src/futures/Ready.h

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#pragma once
+
+#include <mutex>
+#include <optional>
+#include <functional>
+#include <vector>
+
+namespace milvus::futures {
+
+/// @brief Ready is a class that holds a value of type T.
+/// value of Ready can be only set into ready by once,
+/// and allows to register callbacks to be called when the value is ready.
+template <class T>
+class Ready {
+ public:
+    Ready() : is_ready_(false){};
+
+    Ready(const Ready<T>&) = delete;
+
+    Ready(Ready<T>&&) noexcept = default;
+
+    Ready&
+    operator=(const Ready<T>&) = delete;
+
+    Ready&
+    operator=(Ready<T>&&) noexcept = default;
+
+    /// @brief set the value into Ready.
+    void
+    setValue(T&& value) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        value_ = std::move(value);
+        is_ready_ = true;
+        // perform all callbacks which is registered before value is ready.
+        for (auto& callback : callbacks_) {
```

Comment:
callback hould happen without holding the lock?

Reply:
Yes, will fix it soon.